crawl-ref/source/mon-place.cc

```diff
@@ -794,15 +610,19 @@ static bool _valid_monster_generation_location(const mgen_data &mg,
         return false;
     }
 
-    // Check player proximity to avoid band members being placed
-    // close to the player erroneously.
-    // XXX: This is a little redundant with proximity checks in
-    // place_monster.
-    if (mg.proximity == PROX_AWAY_FROM_PLAYER
-        && grid_distance(you.pos(), mg_pos) <= LOS_RADIUS)
+    bool close_to_player = grid_distance(you.pos(), mg_pos) <= LOS_RADIUS;
+    if (mg.proximity == PROX_AWAY_FROM_PLAYER && close_to_player
+        || mg.proximity == PROX_CLOSE_TO_PLAYER && !close_to_player)
     {
         return false;
     }
+    // Check that the location is not proximal to level stairs.
+    else if (mg.proximity == PROX_AWAY_FROM_STAIRS)
+    {
+        for (distance_iterator di(mg_pos, false, LOS_RADIUS); di; ++di)
```

Comment:
Why distance_iterator instead of radius_iterator?

Also, this forbids areas that are on the other side of some solid walls, which are in a sense not actually near any stairs. Probably not a big deal, but if it's easy you might consider something like figuring out the pathfinding distance from a position to the nearest stair.